### ISSUE TITLE
How hard and performant would a "query_ellipse" method be to implement?

### ISSUE THREAD

**Issue by wmwv** [LSSTDESC/hpgeom#10]
Following the "query_circle" implementation:

https://github.com/LSSTDESC/hpgeom/blob/main/hpgeom/healpix_geom.c#L787

```
double cangdist = cosdist_zphi(ptg_z, ptg_phi, pix_z, pix_phi);
[...]
    int zone = (cangdist < cosrad) ? 1 : ((cangdist <= crmdr[o]) ? 2 : 3);
```

where
https://github.com/LSSTDESC/hpgeom/blob/main/hpgeom/healpix_geom.c#L526

```
inline double cosdist_zphi(double z1, double phi1, double z2, double phi2) {
    return z1 * z2 + cos(phi1 - phi2) * sqrt((1. - z1 * z1) * (1. - z2 * z2));
}
```
---

Could a "query_ellipse" method be implemented as something like

```
double cnormdist = cos_normdist_thetaphi(ptg_theta, ptg_phi, pix_theta, pix_phi, rA, rB, pa);
```

```
inline double normdistsq_thetaphi(double theta1, double phi1, double theta2, double phi2, double rA, double rB, double pa) {
    scale_xx = pow(cos(pa) / rA, 2) + pow(sin(pa) / rB, 2);
    scale_yy = pow(sin(pa) / rA), 2) + pow(cos(pa) / rB, 2);
    scale_xy = 2 * cos(pa) * sin(pa) * (pow(1/rA, 2) + pow(1/rB, 2))
    x = cos(theta1 - theta2)
    y = sin(theta1 - theta2)

    R_sq = pow(scale_xx * x, 2) + pow(scale_yy * y, 2) + scale_xy * x * y

    return rA * sqrt(R_sq)
}
```

where rA, rB (and PA) are in radians.  (The above could be rewritten in terms of rA/rB and potential avoid some divisions; but maybe the compiler is smart enough to figure this out anyway).

And then compare to cosrad, crmdr, crpdr as for `query_circle`.

This is effectively redefining `crmdr` and `crpdr` in terms of `rA` instead of `radius`, which must not quite be right.  But spherical geometry makes my brain hurt so I'll delay this until later.

**Comment by wmwv** [comment]
I'm sure the above isn't actually right for significant separations of theta1, theta2.  Need to go brush up on my spherical trig.

**Comment by wmwv** [comment]
My reading of the Q3C code (GPL v2.0) is that it uses polynomial approximations for the ellipses.  It also handles the large and small ellipse case separately, with a note that ellipses larger than 35 degrees breaks assumptions than an ellipse is convex within a Q3C pixel.

https://github.com/segasai/q3c/blob/14e7610febf7577a180ee27408a7a5a24a216121/q3cube.c#L2848

**Comment by erykoff** [comment]
What are the use cases here?  I think we can demand a maximum ellipse size, at least for a first pass.

**Comment by wmwv** [comment]
The use case I have in mind is finding pixels within several elliptical effective radii of a catalog of galaxies.  And agreed, only doing this for things < 1 degree in size would be fine (sorry, M31).

**Comment by erykoff** [comment]
Closed with #17. (To answer the question in the issue subject: “not too hard”.)
